### ISSUE TITLE
Support tuple index with `all_tuples!`

### ISSUE THREAD

**Issue by viridia** [bevyengine/bevy#10915]
## What problem does this solve or what need does it fill?

Other tuple macro crates such as `impl_trait_for_tuples` and `fortuples` let you get the numeric tuple index. For example, in `impl_trait_for_tuples` if you say `self.T` it knows that you want the numeric index `self.0` rather than the type name, whereas if you write `$T` it knows you want the typename `T0`.

This is useful for traits that gather up the tuple members:

```rust
Vec::from([self.0.into(), self.1.into(), self.2.into()])
```

## What solution would you like?

Basically some syntax that would allow you access to the numeric index of the tuple, as opposed to the Nth type name.

## What alternative(s) have you considered?

Continuing to use the third-party crate instead of Bevy's own solution.

**Comment by VVishion** [comment]
Does
```rust
let tuple_items = ($($t,)*);
$( $t.into() )*
```
do the trick?

**Comment by viridia** [comment]
Collecting into a vector was just an example, here's a different one:

The goal in this example is to allow a method to take a variable list of arguments which may not initially be of the same type, but which all implement the same trait and which get converted into the same type:

```rust
element.class_names((
    "dragging",
    "expanded",
    "selected".if_true(is_selected)
 ));
```
Here's how I implemented this with the `impl_for_tuples` crate:

```rust
#[impl_for_tuples(1, 16)]
impl<'a> ClassNamesTuple<'a> for Tuple {
    for_tuples!( where #( Tuple: ClassName )* );

    /// Convert the arguments to a HashSet, discarding elements that have a false condition:
    fn to_set(self) -> HashSet<String> {
        let mut result = HashSet::<String>::new();
        for_tuples!( #(
            if let Some(cls) = self.Tuple.to_class() {
                result.insert(cls.to_owned());
            } )* );
        result
    }
}
```

**Comment by VVishion** [comment]
Something like this?
```rust
use bevy::utils::all_tuples;

macro_rules! impl_class_name {
    ($($t:ident),*) => {
        #[allow(non_snake_case)]
        #[allow(unused_mut)]
        #[allow(unused_variables)]
        impl<'a, $($t),*> ClassNamesTuple<'a> for ($($t,)*)
        {
            fn to_set(self) -> HashSet<String> {
                let mut result = HashSet::<String>::new();
                let ($($t,)*) = self;
                $(if let Some(cls) = $t.to_class() {
                    result.insert(cls.to_owned());
                })*

                result
            }
        }
    };
}

all_tuples!(impl_class_name, 1, 16, C);
```

**Comment by viridia** [comment]
It would be helpful if the documentation for bevy all_tuples had some explanations. I find it rather difficult to look at a bare, uncommented example and figure out what is going on.

**Comment by VVishion** [comment]
all tuples just expands the following
```rust
impl_class_name() // if we start from 0
impl_class_name(C0);
impl_class_name(C0, C1);
impl_class_name(C0, C1, C2);
.
.
.
```

**Comment by VVishion** [comment]
The trick with the tuple index is
```rust
let ($($t,)*) = self;
if let Some(cls) = $t.to_class() {
    result.insert(cls.to_owned());
})*
```
Which expands to the following for lets say `impl_class_name(C0, C1, C2);`
```rust
let (C0, C1, C2) = self;
if let Some(cls) = C0.to_class() {
    result.insert(cls.to_owned());
}
if let Some(cls) = C1.to_class() {
    result.insert(cls.to_owned());
}
if let Some(cls) = C2.to_class() {
    result.insert(cls.to_owned());
}

**Comment by viridia** [comment]
What I need is something that expands to the following:
```rust
impl_effect_tuple!((P0, 0));
impl_effect_tuple!((P0, 0), (P1, 1));
impl_effect_tuple!((P0, 0), (P1, 1), (P2, 2));
```
Either that or:
```rust
impl_effect_tuple!(P0, 0; P1, 1; P2, 2);
```
Either one will work. Xilem has a macro which uses the second form.

With that, I can write macros that look like this:
```rust
macro_rules! impl_effect_tuple {
    ( $($effect: ident, $idx: tt);+ ) => {
        impl<$(
            $effect: EntityEffect,
        )+> EffectTuple for ( $( $effect, )+ ) {
            type State = ( $( $effect::State ),*, );

            fn apply(&self, cx: &mut Cx, target: Entity) -> Self::State {
                ($( self.$idx.apply(cx, target) ),*,)
            }

            fn reapply(&self, cx: &mut Cx, target: Entity, state: &mut Self::State) {
                $( self.$idx.reapply(cx, target, &mut state.$idx); )*
            }
        }
    };
}
```
So for example the `apply` method would expand to:
```rust
fn apply(&self, cx: &mut Cx, target: Entity) -> Self::State {
    (self.0.apply(cx, target),
        self.1.apply(cx, target),
        self.2.apply(cx, target),)
}
```
In other words, it allows a way to transform/map varargs tuples.

**Comment by alice-i-cecile** [comment]
Implemented in `variadics-plz` now :)
